pkg/deps/collate.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deps
+
+import (
+	"fmt"
+	"sort"
+
+	bundle "github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/apis/bundle/v1alpha1"
+	"github.com/blang/semver"
+)
+
+// searchOpts are options for searching for latest or previous versions. For a
+// more detailed discussion, see the ResolveOptions.
+type searchOpts struct {
+	match          map[string][]string
+	matchIfPresent map[string][]string
+	exclude        map[string][]string
+}
+
+// String returns the stringified search options.
+func (s *searchOpts) String() string {
+	return fmt.Sprintf("{%+v, %+v}", s.match, s.exclude)
+}
+
+// sortedVersions contain all the component versions for a single component,
+// sorted by version. It should be considered immutable and must not be changed
+// once created.
+type sortedVersions struct {
+	componentName string
+
+	// versions contains the versions sorted in ascending order. In otherwords,
+	// versions[0] is the lowest version.
+	versions []*depMeta
+}
+
+// String prints the stringified components.
+func (s *sortedVersions) String() string {
+	out := fmt.Sprintf("{%s", s.componentName)
+	for i, v := range s.versions {
+		out += " "
+		if i == 0 {
+			out += "["
+		}
+		out += v.version.String()
+		if i == len(s.versions)-1 {
+			out += "]"
+		}
+	}
+	return out + "}"
+}
+
+// useComponent evaluates whether, based on the searchOpts, the component
+// should be used.
+func (s *sortedVersions) useComponent(m *depMeta, opts *searchOpts) bool {
+	if opts == nil {
+		return true
+	}
+
+	match := true
+	for k, vals := range opts.match {
+		matchesOne := false
+		for i := range vals {
+			if m.annotations[k] == vals[i] {
```

Comment:
What happens if m.annotations is length 1 and opts.match is length 5? 
Seems like that errors, why is that invalid inputs?

Reply:
Oh, good point. this is wrong. It should be matching all